File: src/engine/Path.ts
export class Path {
  public static parentId = "^";

  public _isRelative: boolean;
  public _components: Path.Component[];
  public _componentsString: string | null;

  constructor();
  constructor(componentsString: string);
  constructor(head: Path.Component, tail: Path);
  constructor(head: Path.Component[], relative?: boolean);
  constructor() {
    this._components = [];
    this._componentsString = null;
    this._isRelative = false;

    if (typeof arguments[0] == "string") {
      let componentsString = arguments[0] as string;
      this.componentsString = componentsString;
    } else if (
      arguments[0] instanceof Path.Component &&
      arguments[1] instanceof Path
    ) {
      let head = arguments[0] as Path.Component;
      let tail = arguments[1] as Path;
      this._components.push(head);
      this._components = this._components.concat(tail._components);
    } else if (arguments[0] instanceof Array) {
      let head = arguments[0] as Path.Component[];
      let relative = !!arguments[1] as boolean;
      this._components = this._components.concat(head);
      this._isRelative = relative;
    }
  }
  get isRelative() {
    return this._isRelative;
  }
  get componentCount(): number {
    return this._components.length;
  }
  get head(): Path.Component | null {
    if (this._components.length > 0) {
      return this._components[0];
    } else {
      return null;
    }
  }
  get tail(): Path {
    if (this._components.length >= 2) {
      // careful, the original code uses length-1 here. This is because the second argument of
      // List.GetRange is a number of elements to extract, wherease Array.slice uses an index
      let tailComps = this._components.slice(1, this._components.length);
      return new Path(tailComps);
    } else {
      return Path.self;
    }
  }
  get length(): number {
    return this._components.length;
  }
  get lastComponent(): Path.Component | null {
    let lastComponentIdx = this._components.length - 1;
    if (lastComponentIdx >= 0) {
      return this._components[lastComponentIdx];
    } else {
      return null;
    }
  }
  get containsNamedComponent(): boolean {
    for (let i = 0, l = this._components.length; i < l; i++) {
      if (!this._components[i].isIndex) {
        return true;
      }
    }
    return false;
  }
  static get self(): Path {
    let path = new Path();
    path._isRelative = true;
    return path;
  }

  public GetComponent(index: number): Path.Component {
    return this._components[index];
  }
  public PathByAppendingPath(pathToAppend: Path): Path {
    let p = new Path();

    let upwardMoves = 0;
    for (let i = 0; i < pathToAppend._components.length; ++i) {
      if (pathToAppend._components[i].isParent) {
        upwardMoves++;
      } else {
        break;
      }
    }

    for (let i = 0; i < this._components.length - upwardMoves; ++i) {
      p._components.push(this._components[i]);
    }

    for (let i = upwardMoves; i < pathToAppend._components.length; ++i) {
      p._components.push(pathToAppend._components[i]);
    }

    return p;
  }
  get componentsString(): string {
    if (this._componentsString == null) {
      this._componentsString = this._components.join(".");
      if (this.isRelative)
        this._componentsString = "." + this._componentsString;
    }

    return this._componentsString;
  }
  set componentsString(value: string) {
    this._components.length = 0;

    this._componentsString = value;

    if (this._componentsString == null || this._componentsString == "") return;

    if (this._componentsString[0] == ".") {
      this._isRelative = true;
      this._componentsString = this._componentsString.substring(1);
    }

    let componentStrings = this._componentsString.split(".");
    for (let str of componentStrings) {
      // we need to distinguish between named components that start with a number, eg "42somewhere", and indexed components
      // the normal parseInt won't do for the detection because it's too relaxed.
      // see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt
      if (/^(\-|\+)?([0-9]+|Infinity)$/.test(str)) {
        this._components.push(new Path.Component(parseInt(str)));
      } else {
        this._components.push(new Path.Component(str));
      }
    }
  }
  public toString(): string {
    return this.componentsString;
  }
  public Equals(otherPath: Path | null): boolean {
    if (otherPath == null) return false;

    if (otherPath._components.length != this._components.length) return false;

    if (otherPath.isRelative != this.isRelative) return false;

    // the original code uses SequenceEqual here, so we need to iterate over the components manually.
    for (let i = 0, l = otherPath._components.length; i < l; i++) {
      // it's not quite clear whether this test should use Equals or a simple == operator,
      // see https://github.com/y-lohse/inkjs/issues/22
      if (!otherPath._components[i].Equals(this._components[i])) return false;
    }

    return true;
  }
  public PathByAppendingComponent(c: Path.Component): Path {
    let p = new Path();
    p._components.push(...this._components);
    p._components.push(c);
    return p;
  }
}

export namespace Path {
  export class Component {
    public readonly index: number;
    public readonly name: string | null;

    constructor(indexOrName: string | number) {
      this.index = -1;
      this.name = null;
      if (typeof indexOrName == "string") {
        this.name = indexOrName;
      } else {
        this.index = indexOrName;
      }
    }
    get isIndex(): boolean {
      return this.index >= 0;
    }
    get isParent(): boolean {
      return this.name == Path.parentId;
    }

    public static ToParent(): Component {
      return new Component(Path.parentId);
    }
    public toString(): string | null {
      if (this.isIndex) {
        return this.index.toString();
      } else {
        return this.name;
      }
    }
    public Equals(otherComp: Component): boolean {
      if (otherComp != null && otherComp.isIndex == this.isIndex) {
        if (this.isIndex) {
          return this.index == otherComp.index;
        } else {
          return this.name == otherComp.name;
        }
      }

      return false;
    }
  }
}

File: ink_dart/lib/src/engine/path.dart
/// Path represents a location within the story content hierarchy.
///
/// Paths can be absolute (from root) or relative (from current position).
/// Components can be either named (string) or indexed (int).
///
/// Examples:
/// - "knot.stitch.0" - absolute path to first element of stitch
/// - ".^.sibling" - relative path: go up then to sibling
class Path {
  static const String parentId = '^';

  final List<PathComponent> _components = [];
  bool _isRelative = false;
  String? _componentsString;

  /// Create an empty path.
  Path();

  /// Create a path from a string representation.
  ///
  /// Format: "component.component.component"
  /// Relative paths start with "."
  /// Parent references use "^"
  Path.fromString(String componentsString) {
    this.componentsString = componentsString;
  }

  /// Create a path from components.
  Path.fromComponents(List<PathComponent> components, {bool relative = false}) {
    _components.addAll(components);
    _isRelative = relative;
  }

  /// Create a path by prepending a head component to a tail path.
  Path.fromHeadTail(PathComponent head, Path tail) {
    _components.add(head);
    _components.addAll(tail._components);
  }

  /// Whether this is a relative path (starts with .)
  bool get isRelative => _isRelative;

  /// Number of components in the path.
  int get componentCount => _components.length;

  /// Alias for componentCount.
  int get length => _components.length;

  /// First component of the path, or null if empty.
  PathComponent? get head {
    if (_components.isNotEmpty) {
      return _components[0];
    }
    return null;
  }

  /// Path without the first component.
  Path get tail {
    if (_components.length >= 2) {
      return Path.fromComponents(_components.sublist(1));
    }
    return Path.self;
  }

  /// Last component of the path, or null if empty.
  PathComponent? get lastComponent {
    if (_components.isNotEmpty) {
      return _components.last;
    }
    return null;
  }

  /// Whether any component is a named (string) component.
  bool get containsNamedComponent {
    for (final comp in _components) {
      if (!comp.isIndex) {
        return true;
      }
    }
    return false;
  }

  /// A relative empty path (represents "self").
  static Path get self {
    final path = Path();
    path._isRelative = true;
    return path;
  }

  /// Get component at the given index.
  PathComponent getComponent(int index) => _components[index];

  /// Create a new path by appending another path.
  ///
  /// Handles parent references (^) by moving up the hierarchy.
  Path pathByAppendingPath(Path pathToAppend) {
    final p = Path();

    int upwardMoves = 0;
    for (int i = 0; i < pathToAppend._components.length; ++i) {
      if (pathToAppend._components[i].isParent) {
        upwardMoves++;
      } else {
        break;
      }
    }

    for (int i = 0; i < _components.length - upwardMoves; ++i) {
      p._components.add(_components[i]);
    }

    for (int i = upwardMoves; i < pathToAppend._components.length; ++i) {
      p._components.add(pathToAppend._components[i]);
    }

    return p;
  }

  /// Create a new path by appending a single component.
  Path pathByAppendingComponent(PathComponent c) {
    final p = Path();
    p._components.addAll(_components);
    p._components.add(c);
    return p;
  }

  /// String representation of the path.
  String get componentsString {
    if (_componentsString == null) {
      _componentsString = _components.join('.');
      if (isRelative) {
        _componentsString = '.$_componentsString';
      }
    }
    return _componentsString!;
  }

  /// Parse a string into path components.
  set componentsString(String value) {
    _components.clear();
    _componentsString = value;

    if (_componentsString == null || _componentsString!.isEmpty) return;

    String parseStr = _componentsString!;

    if (parseStr[0] == '.') {
      _isRelative = true;
      parseStr = parseStr.substring(1);
    }

    if (parseStr.isEmpty) return;

    final componentStrings = parseStr.split('.');
    for (final str in componentStrings) {
      // Check if it's a pure integer (not a named component starting with digits)
      if (RegExp(r'^(-|\+)?([0-9]+|Infinity)$').hasMatch(str)) {
        _components.add(PathComponent.fromIndex(int.parse(str)));
      } else {
        _components.add(PathComponent.fromName(str));
      }
    }
  }

  @override
  String toString() => componentsString;

  /// Check equality with another path.
  bool equals(Path? otherPath) {
    if (otherPath == null) return false;
    if (otherPath._components.length != _components.length) return false;
    if (otherPath.isRelative != isRelative) return false;

    for (int i = 0; i < _components.length; i++) {
      if (!_components[i].equals(otherPath._components[i])) {
        return false;
      }
    }

    return true;
  }

  @override
  bool operator ==(Object other) {
    if (other is Path) {
      return equals(other);
    }
    return false;
  }

  @override
  int get hashCode => componentsString.hashCode;
}

/// A single component of a Path.
///
/// Can be either:
/// - An index (int >= 0) for positional access
/// - A name (string) for named content access
/// - Parent reference (^) for going up the hierarchy
class PathComponent {
  final int index;
  final String? name;

  /// Create a component from an index.
  PathComponent.fromIndex(this.index) : name = null;

  /// Create a component from a name.
  PathComponent.fromName(this.name) : index = -1;

  /// Create a parent reference component.
  factory PathComponent.toParent() => PathComponent.fromName(Path.parentId);

  /// Whether this component is an index (vs a name).
  bool get isIndex => index >= 0;

  /// Whether this component is a parent reference (^).
  bool get isParent => name == Path.parentId;

  @override
  String toString() {
    if (isIndex) {
      return index.toString();
    }
    return name ?? '';
  }

  /// Check equality with another component.
  bool equals(PathComponent other) {
    if (isIndex == other.isIndex) {
      if (isIndex) {
        return index == other.index;
      } else {
        return name == other.name;
      }
    }
    return false;
  }

  @override
  bool operator ==(Object other) {
    if (other is PathComponent) {
      return equals(other);
    }
    return false;
  }

  @override
  int get hashCode => isIndex ? index.hashCode : (name?.hashCode ?? 0);
}
